src/main/scala/io/vlach/CheckoutSystem.scala

```diff
@@ -5,34 +5,17 @@ case object Apple extends Fruit(cost = 60L)
 case object Orange extends Fruit(cost = 25L)
 
 object CheckoutSystem {
-  def checkout(items: List[Fruit]): Long = {
-    val (applesUntyped, orangesUntyped) = items.partition {
-      case Apple => true
-      case Orange => false
-    }
-    val apples = applesUntyped.collect { case Apple => Apple }
-    val oranges = orangesUntyped.collect { case Orange => Orange }
-
-    val applesToCheckout = twoForOne(apples)
-    val orangesToCheckout = threeForTwo(oranges)
 
-    (applesToCheckout ++ orangesToCheckout).foldRight(0L)((fruit, acc) => fruit.cost + acc)
+  def checkout(items: List[Fruit]): Long = {
+    val (apples, oranges) = items.partition(_ == Apple)
+    (twoForOne(apples) ++ threeForTwo(oranges)).map(_.cost).sum
   }
 
-  private def twoForOne(apples: List[Apple.type]): List[Apple.type] = {
```

Comment:
don't understand the rationale behind Apple.type here but suspect there may have been good reason of this